### DISCUSSION TITLE
How to port iPXE to custom PCIe network card

### DISCUSSION THREAD

**Discussion by Kostr** [ipxe/ipxe#1019] [Q&A]
I haven't found any instructions about how to add support for a new PCIe network card to iPXE.

Is it just a matter of writing a driver similar to the ones present in the `src/drivers/net` folder? Or is there something more?

Are there any guides/helpers for this task besides the skeleton driver?

https://github.com/ipxe/ipxe/blob/master/src/drivers/net/skeleton.c

https://github.com/ipxe/ipxe/blob/master/src/drivers/net/skeleton.h

**Comment by mcb30** [comment]
Yes, it's a case of writing a driver similar to the ones present in `src/drivers/net`.

You will usually want to start from the network card datasheet and add only the definitions and code that are actually required in iPXE's very simple network card driver model.  You can often use the corresponding Linux driver as a reference to resolve ambiguities in the datasheet.  It is almost never a good idea to try "porting" the Linux driver to iPXE, since this is generally much more work than creating a new driver from scratch.

**Reply by Kostr** [reply]
> add only the definitions and code that are actually required in iPXE's very simple network card driver model

What are those?

From the other drivers code it looks like I need to populate `struct net_device_operations` with these functions at least 
```
  open
  close
  transmit
  poll
```
I.e everything without irq.

But can I start with a more simple implementation? For example `net_device_operations` with just `open/close`. What tests can I perform in this case?

Also what images should I create for the tests? The`.efi` application image or a driver image `.efidrv` that I would test with standard EDK2 network stack?

I guess I'm looking for some iterative approach to driver write. Steps with features that I need to implement and some hints how to test every step.

**Reply by NiKiZe** [reply]
There is https://ipxe.org/dev/driver which might be a good start
